### ISSUE TITLE
an illegal memory access when parallel training 

### ISSUE THREAD

**Issue by Jinwook-Kim** [Microsoft/CNTK#367]
When I run ResNet_152 (an ImageNet training example) parallel training example with 2 or 3 TitanX GPUs,  an illegal memory access error is occured.
For the first batches (about 18,000 images) the parallel training (2~3 GPUs) works well, but after that, the process dies showing the error message. 
A single GPU training has no problem so far. (training for 5~6 days)

Would anyone give me some suggestion for solving this issue, please?

The following is my detailed execution command and the error messages. 

* OS: Ubuntu 14.04
* GPU: Nvidia Titan X
* ConfigFile: 
`Examples/Image/Miscellaneous/ImageNet/ResNet/ResNet_152.cntk`
(I just reduced the training/ testing mini-batch size for avoiding an out of memory error)

* Execution command: 
`mpiexec -n 2 cntk configFile=ResNet_152.cntk parallelTrain=true deviceId=auto`
(`-n 2` and `-n 3` both makes the same problem, but `-n 1` works good)

* Error message (only for logrank1 file): 
```
779 out of 1296 nodes do not share the minibatch layout with the input data.

Post-processing network complete.

Loaded model with 1296 nodes on GPU 2.

Training criterion node(s):
	CE = CrossEntropyWithSoftmax

Evaluation criterion node(s):

	Err = ErrorPrediction


Allocating matrices for forward and/or backward propagation.
No PreCompute nodes found, skipping PreCompute step.

Starting Epoch 1: learning rate per sample = 0.022222  effective momentum = 0.900000  momentum as time constant = 427.1 samples

Starting minibatch loop, DataParallelSGD training (MyRank = 1, NumNodes = 3, NumGradientBits = 32), distributed reading is ENABLED.


 Epoch[ 1 of 125]-Minibatch[   1- 500]: SamplesSeen = 22500; TrainLossPerSample =  7.07490356; EvalErr[0]PerSample = 0.99884444; TotalTime = 511.3217s; SamplesPerSecond = 44.0
 Epoch[ 1 of 125]-Minibatch[ 501-1000]: SamplesSeen = 22500; TrainLossPerSample =  6.88385591; EvalErr[0]PerSample = 0.99857778; TotalTime = 566.9095s; SamplesPerSecond = 39.7
 Epoch[ 1 of 125]-Minibatch[1001-1500]: SamplesSeen = 22500; TrainLossPerSample =  6.86290641; EvalErr[0]PerSample = 0.99831111; TotalTime = 607.1935s; SamplesPerSecond = 37.1
 Epoch[ 1 of 125]-Minibatch[1501-2000]: SamplesSeen = 22500; TrainLossPerSample =  6.84834738; EvalErr[0]PerSample = 0.99764444; TotalTime = 528.5081s; SamplesPerSecond = 42.6
 Epoch[ 1 of 125]-Minibatch[2001-2500]: SamplesSeen = 22500; TrainLossPerSample =  6.82727752; EvalErr[0]PerSample = 0.99768889; TotalTime = 593.8823s; SamplesPerSecond = 37.9
 Epoch[ 1 of 125]-Minibatch[2501-3000]: SamplesSeen = 22500; TrainLossPerSample =  6.80221521; EvalErr[0]PerSample = 0.99733333; TotalTime = 600.8662s; SamplesPerSecond = 37.4
 Epoch[ 1 of 125]-Minibatch[3001-3500]: SamplesSeen = 22500; TrainLossPerSample =  6.78126761; EvalErr[0]PerSample = 0.99640000; TotalTime = 558.1656s; SamplesPerSecond = 40.3
 Epoch[ 1 of 125]-Minibatch[3501-4000]: SamplesSeen = 22500; TrainLossPerSample =  6.75373040; EvalErr[0]PerSample = 0.99688889; TotalTime = 478.0318s; SamplesPerSecond = 47.1
 Epoch[ 1 of 125]-Minibatch[4001-4500]: SamplesSeen = 22500; TrainLossPerSample =  6.72999766; EvalErr[0]PerSample = 0.99715556; TotalTime = 509.2051s; SamplesPerSecond = 44.2
 Epoch[ 1 of 125]-Minibatch[4501-5000]: SamplesSeen = 22500; TrainLossPerSample =  6.68511361; EvalErr[0]PerSample = 0.99560000; TotalTime = 514.9840s; SamplesPerSecond = 43.7
 Epoch[ 1 of 125]-Minibatch[5001-5500]: SamplesSeen = 22500; TrainLossPerSample =  6.64548318; EvalErr[0]PerSample = 0.99560000; TotalTime = 500.8263s; SamplesPerSecond = 44.9
 Epoch[ 1 of 125]-Minibatch[5501-6000]: SamplesSeen = 22500; TrainLossPerSample =  6.60723752; EvalErr[0]PerSample = 0.99426667; TotalTime = 478.9874s; SamplesPerSecond = 47.0
 Epoch[ 1 of 125]-Minibatch[6001-6500]: SamplesSeen = 22500; TrainLossPerSample =  6.54260951; EvalErr[0]PerSample = 0.99297778; TotalTime = 500.1494s; SamplesPerSecond = 45.0
 Epoch[ 1 of 125]-Minibatch[6501-7000]: SamplesSeen = 22500; TrainLossPerSample =  6.48370495; EvalErr[0]PerSample = 0.99186667; TotalTime = 501.8939s; SamplesPerSecond = 44.8
 Epoch[ 1 of 125]-Minibatch[7001-7500]: SamplesSeen = 22500; TrainLossPerSample =  6.43145192; EvalErr[0]PerSample = 0.99066667; TotalTime = 484.0805s; SamplesPerSecond = 46.5
 Epoch[ 1 of 125]-Minibatch[7501-8000]: SamplesSeen = 22500; TrainLossPerSample =  6.35213712; EvalErr[0]PerSample = 0.98880000; TotalTime = 505.0759s; SamplesPerSecond = 44.5
 Epoch[ 1 of 125]-Minibatch[8001-8500]: SamplesSeen = 22500; TrainLossPerSample =  6.27174462; EvalErr[0]PerSample = 0.98675556; TotalTime = 499.6164s; SamplesPerSecond = 45.0
 Epoch[ 1 of 125]-Minibatch[8501-9000]: SamplesSeen = 22500; TrainLossPerSample =  6.19746436; EvalErr[0]PerSample = 0.98480000; TotalTime = 486.1613s; SamplesPerSecond = 46.3
 Epoch[ 1 of 125]-Minibatch[9001-9500]: SamplesSeen = 22500; TrainLossPerSample =  6.14027952; EvalErr[0]PerSample = 0.98128889; TotalTime = 494.5784s; SamplesPerSecond = 45.5
 Epoch[ 1 of 125]-Minibatch[9501-10000]: SamplesSeen = 22500; TrainLossPerSample =  6.08099082; EvalErr[0]PerSample = 0.98035556; TotalTime = 494.4652s; SamplesPerSecond = 45.5
 Epoch[ 1 of 125]-Minibatch[10001-10500]: SamplesSeen = 22500; TrainLossPerSample =  5.99841505; EvalErr[0]PerSample = 0.97702222; TotalTime = 481.2851s; SamplesPerSecond = 46.7
 Epoch[ 1 of 125]-Minibatch[10501-11000]: SamplesSeen = 22500; TrainLossPerSample =  5.93937506; EvalErr[0]PerSample = 0.97422222; TotalTime = 496.7225s; SamplesPerSecond = 45.3
 Epoch[ 1 of 125]-Minibatch[11001-11500]: SamplesSeen = 22500; TrainLossPerSample =  5.88052229; EvalErr[0]PerSample = 0.97204444; TotalTime = 482.8862s; SamplesPerSecond = 46.6
 Epoch[ 1 of 125]-Minibatch[11501-12000]: SamplesSeen = 22500; TrainLossPerSample =  5.82689628; EvalErr[0]PerSample = 0.96911111; TotalTime = 494.8632s; SamplesPerSecond = 45.5
 Epoch[ 1 of 125]-Minibatch[12001-12500]: SamplesSeen = 22500; TrainLossPerSample =  5.78526479; EvalErr[0]PerSample = 0.96728889; TotalTime = 489.2012s; SamplesPerSecond = 46.0
 Epoch[ 1 of 125]-Minibatch[12501-13000]: SamplesSeen = 22500; TrainLossPerSample =  5.72365272; EvalErr[0]PerSample = 0.96208889; TotalTime = 493.5552s; SamplesPerSecond = 45.6
 Epoch[ 1 of 125]-Minibatch[13001-13500]: SamplesSeen = 22500; TrainLossPerSample =  5.66468574; EvalErr[0]PerSample = 0.95933333; TotalTime = 482.7917s; SamplesPerSecond = 46.6
 Epoch[ 1 of 125]-Minibatch[13501-14000]: SamplesSeen = 22500; TrainLossPerSample =  5.61126183; EvalErr[0]PerSample = 0.95724444; TotalTime = 483.9569s; SamplesPerSecond = 46.5
 Epoch[ 1 of 125]-Minibatch[14001-14500]: SamplesSeen = 22500; TrainLossPerSample =  5.54847293; EvalErr[0]PerSample = 0.95235556; TotalTime = 501.7715s; SamplesPerSecond = 44.8
 Epoch[ 1 of 125]-Minibatch[14501-15000]: SamplesSeen = 22500; TrainLossPerSample =  5.51270139; EvalErr[0]PerSample = 0.94591111; TotalTime = 492.0935s; SamplesPerSecond = 45.7
 Epoch[ 1 of 125]-Minibatch[15001-15500]: SamplesSeen = 22500; TrainLossPerSample =  5.46947769; EvalErr[0]PerSample = 0.94751111; TotalTime = 491.1623s; SamplesPerSecond = 45.8
 Epoch[ 1 of 125]-Minibatch[15501-16000]: SamplesSeen = 22500; TrainLossPerSample =  5.40878874; EvalErr[0]PerSample = 0.94040000; TotalTime = 494.2490s; SamplesPerSecond = 45.5
 Epoch[ 1 of 125]-Minibatch[16001-16500]: SamplesSeen = 22500; TrainLossPerSample =  5.37302377; EvalErr[0]PerSample = 0.93742222; TotalTime = 481.5105s; SamplesPerSecond = 46.7
 Epoch[ 1 of 125]-Minibatch[16501-17000]: SamplesSeen = 22500; TrainLossPerSample =  5.32060593; EvalErr[0]PerSample = 0.93217778; TotalTime = 486.4217s; SamplesPerSecond = 46.3
 Epoch[ 1 of 125]-Minibatch[17001-17500]: SamplesSeen = 22500; TrainLossPerSample =  5.29912303; EvalErr[0]PerSample = 0.93093333; TotalTime = 503.9877s; SamplesPerSecond = 44.6
 Epoch[ 1 of 125]-Minibatch[17501-18000]: SamplesSeen = 22500; TrainLossPerSample =  5.25361448; EvalErr[0]PerSample = 0.92924444; TotalTime = 545.6036s; SamplesPerSecond = 41.2
 Epoch[ 1 of 125]-Minibatch[18001-18500]: SamplesSeen = 22500; TrainLossPerSample =  5.23281561; EvalErr[0]PerSample = 0.92724444; TotalTime = 509.0296s; SamplesPerSecond = 44.2
CUDA failure 77: an illegal memory access was encountered ; GPU=2 ; hostname=neptune02 ; expr=cudaMemset(Data(), 0, N * sizeof(ElemType))
pure virtual method called
terminate called without an active exception
[neptune02:32563] *** Process received signal ***
[neptune02:32563] Signal: Aborted (6)
[neptune02:32563] Signal code:  (-6)
[neptune02:32563] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x10340)[0x7f2f7f2ea340]
[neptune02:32563] [ 1] /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x39)[0x7f2f78368cc9]
[neptune02:32563] [ 2] /lib/x86_64-linux-gnu/libc.so.6(abort+0x148)[0x7f2f7836c0d8]
[neptune02:32563] [ 3] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(_ZN9__gnu_cxx27__verbose_terminate_handlerEv+0x155)[0x7f2f78c73535]
[neptune02:32563] [ 4] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x5e6d6)[0x7f2f78c716d6]
[neptune02:32563] [ 5] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x5e703)[0x7f2f78c71703]
[neptune02:32563] [ 6] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x5f1bf)[0x7f2f78c721bf]
[neptune02:32563] [ 7] /home/neptune/cntk/build/release/bin/../lib/libcntkmath.so(+0x361209)[0x7f2f79766209]
[neptune02:32563] [ 8] /opt/acml5.3.1/ifort64_mp/lib/libiomp5.so(+0x50bc9)[0x7f2f6949bbc9]
[neptune02:32563] [ 9] /opt/acml5.3.1/ifort64_mp/lib/libiomp5.so(__kmp_invoke_microtask+0x93)[0x7f2f694ccfe3]
[neptune02:32563] *** End of error message ***
```

**Comment by Alexey-Kamenev** [comment]
Hello,
That's interesting, I cannot repro it on our K40 machines but let me try something else. 
Have you tried reducing `minibatchSize` to some very small values like 4 or 8?
Also, does it happen only in ResNet_152 or with other configs as well? Can you please try running ResNet_50/34/18 and see if you get this error?

**Comment by zruri** [comment]
Hello, 
Thanks for your reply.
So far, I've tried to decrease `minibatchSize=20` for both train and test (that means 10 for each GPU, am I right?), but, I remember, I saw a similar error. (I'm gonna check it again)

For multi-GPU training, I have not tried other configs. (For single-GPU training, I've run ResNet_50)
I will try ResNet_152/50/34/18 with much smaller `minibatchSize` as you suggested and let you know. :)
